src/main/groovy/wooga/gradle/paket/unity/tasks/PaketUnwrapUPMPackages.groovy

```diff
@@ -0,0 +1,199 @@
+    @OutputDirectory
+    File getOutputDirectory() {
+        new File(getReferencesFile().getParentFile(), "Packages")
+    }
+
+    /**
+     * @return the paket.upm.wrapper.reference's of the wrapper packages to unwrap into the Unity3D/Packages project.
+     */
+    @InputFiles
+    FileCollection getInputFiles() {
+        Set<File> files = []
+        def references = new PaketUnityReferences(getReferencesFile())
+
+        if (!getLockFile().exists()) {
+            return null
+        }
+
+        def locks = new PaketLock(getLockFile())
+        def dependencies = locks.getAllDependencies(references.nugets)
+        dependencies.each { nuget ->
+            def depFiles = getFilesForPackage(nuget)
+            files << depFiles
+        }
+
+        project.files(files)
+    }
+
+    Set<File> getFilesForPackage(String nuget) {
+        if (isUPMWrapper(nuget)) {
+            return [new PaketUPMWrapperReference(nuget, project).file] as Set<File>
+        }
+    }
+
+    PaketUnwrapUPMPackages() {
+        description = 'Unwraps wrapped UPM into paket dependencies into unity projects as local UPM Overrides'
+        group = PaketUnityPlugin.GROUP
+    }
+
+    @TaskAction
+    protected performCopy(IncrementalTaskInputs inputs) {
+        if (!inputs.incremental) {
+            cleanLocalUpmOverrides()
+        }
+
+        inputs.outOfDate(new Action<InputFileDetails>() {
+            @Override
+            void execute(InputFileDetails outOfDate) {
+                if(inputFiles.contains(outOfDate.file) && PaketUPMWrapperReference.IsReferenceFile(outOfDate.file)) {
+                    def upmWrapperRef = new PaketUPMWrapperReference(outOfDate.file)
+                    if (upmWrapperRef.exists) {
+                        cleanLocalUpmOverride(upmWrapperRef)
+                        def upmUnwrapper = upmWrapperRef.upmPackageURL.startsWith("http")
+                                ? new DownloadAndUnpackTar(upmWrapperRef, getOutputDirectory())
+                                : new UnwrapUpm(upmWrapperRef, getOutputDirectory())
+                        upmUnwrapper.exec(project)
+                    }
+                }
+            }
+        })
+
+        inputs.removed(new Action<InputFileDetails>() {
+            @Override
+            void execute(InputFileDetails removed) {
+                if (PaketUPMWrapperReference.IsReferenceFile(removed.file)) {
+                    logger.info("remove: ${removed.file}")
+                    def upmWrapper = new PaketUPMWrapperReference(removed.file);
+                    if (upmWrapper.exists) {
+                        cleanLocalUpmOverride(new PaketUPMWrapperReference(removed.file))
+                    }
+                    removed.file.delete()
+                    assert removed.removed
+                }
+            }
+        })
+    }
+
+    public static boolean isUPMWrapper(String packageName) {
+        return packageName.startsWith(localUPMWrapperPackagePrefix)
+    }
```

Comment:
I wonder if we actually need this. The fact that a reference file is in the package is already enough or? I mean that is also the way the Unity install tasks checks of the package needs to be skipped.

Reply:
I also thought about this a bit. I hit some roadbumps with issues when deleting existing files/references but i managed to overcome them now, i think i can refactor this out. Then a wrapper is just a WDK which has this reference file in it.